### ISSUE TITLE
Terminating folds/parsers should return Step type in "initial"

### ISSUE THREAD

**Issue by harendra-kumar** [composewell/streamly#671]
Consider the following cases:

* `take 0`, the fold cannot terminate until it is fed at least one element. Here is a case in Linux fs event handling code where it is problematic:

```
readOneEvent wdMap = do
    let headerLen = (sizeOf (undefined :: CInt)) + 12
    arr <- PR.take headerLen (A.writeN headerLen)
    (ewd, eflags, cookie, pathLen) <- PR.yieldM $ A.asPtr arr readHeader
    -- XXX need the "initial" in parsers to return a step type so that "take 0"
    -- can return without an input. otherwise if pathLen is 0 we will keep
    -- waiting to read one more char before we return this event.
    path <-
        if pathLen /= 0
        then do
            -- XXX sliceSepByMax drops the separator so assumes a null
            -- terminated path, we should use a takeWhile nested inside a
            -- takeP
            pth <- PR.sliceSepByMax (== 0) pathLen (A.writeN pathLen)
            let remaining = pathLen - A.length pth - 1
            when (remaining /= 0) $ PR.take remaining FL.drain
            return pth
        else return $ A.fromList []
```

* Also when we use the `sliceEndWith` parser with `parseMany` we cannot return an empty output if the input is empty because we always call "initial" and it always returns a default value of the fold. Though we can fix this by modifying `parseMany` to call `initial` when it sees first element in the stream instead of calling it unconditionally. But it means the fold driver (i.e. `parseMany`) has to be written differently for different cases. The choice of whether to return a default value or nothing should be left to the fold and not the driver.

In case of parsers we may want to return Error immediately in `initial` instead of waiting for an element to be consumed. That requires `initial` to return Error.

**Comment by harendra-kumar** [comment]
For folds we can fix this in 0.8.0 itself so that we do not have two implementations of `foldMany`. And it should not be too difficult for folds. For parsers we can do it in 0.8.1.

**Comment by harendra-kumar** [comment]
The `yield` of `ParserD` also has the same issue, it would not yield a value unless an input is fed to the parser.

**Comment by harendra-kumar** [comment]
This should be partially fixed #787. However, we still need to fix the parsers. For parsers we can have a limited fix for now, basically using only `Partial`, `Done` and `Error` in initial and not supporting `Continue`.

**Comment by harendra-kumar** [comment]
To keep the parser implementation simpler and not have any partial functions we could use separate types for `initial` and `extract` return types instead of reusing the `Step` type:

```
data Initial s b =
      IPartial s
    | IDone b  
    | IError String

data Extract s b =
      EPartial Int b -- partial result and unused input count
    | EDone Int b   -- final result with unused input count
    | EError String
```

These are essentially subtypes of the `Step` type. They both omit `Continue` for backtracking.`Initial` does not allow any backtracking at all, even in `Partial` and `Done` because we have not consumed any input yet.
